### PR TITLE
Fix TopK MoE router op fusion failure

### PR DESCRIPTION
**Pull request by yosh20004**
## Overview

This PR fixes a TopK MoE router fusion failure caused by overly conservative
overlap detection in `ggml_cuda_check_fusion_memory_ranges()`.

In the TopK MoE router path, some tensors may intentionally alias the router
logits buffer. The existing fusion memory range check treats this as a possible
unsafe overwrite and rejects the fusion, even when the overlap is a known case
that can be handled safely by the fused op.

This PR makes two changes:

- Adds a whitelist mechanism to `ggml_cuda_check_fusion_memory_ranges()` so that
  selected, known-safe overlap cases can be handled by the fused CUDA op instead
  of causing fusion to be rejected unconditionally.
- Adds a `weights_alias_logits` path to `ggml_cuda_op_topk_moe` to make the
  aliased-weights case safe at runtime.

## Additional information

- Test device: A100-40G
- Test model: Qwen3-30B-A3B-UD-IQ1_S.gguf
- Test command:
```
./build/bin/llama-bench \
  -m ./models/unsloth-Qwen3-30B-A3B-GGUF/Qwen3-30B-A3B-UD-IQ1_S.gguf \
  -ngl 99 \
  -p 1024,2048,4096,8192 \
  -n 0 \
  -b 2048 \
  -ub 512 \
  -r 20 \
  -o json
```
- Result:

| Test | Samples | Master t/s | Fix t/s | Speedup |
|---|---:|---:|---:|---:|
| pp1024 | 20 | 2946.28 | 2983.94 | +1.28% |
| pp2048 | 20 | 2850.62 | 2901.61 | +1.79% |
| pp4096 | 20 | 2662.66 | 2694.76 | +1.21% |
| pp8192 | 10 | 2293.99 | 2322.65 | +1.25% |

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES  I used AI assistance and reviewed the generated code myself.

### CODE REVIEW

**Comment by am17an** [comment]
Not sure why you think it's safe that the weights can alias logits. Weights are definitely not read in their entirety before writing logits when nrows > 1

**Comment by yosh20004** [comment]
> Weights are definitely not read in their entirety before writing logits when nrows > 1

Thanks for the review. I agree that weights aliasing logits is not safe when nrows > 1. To avoid that case, I added a temporary buffer path in ggml_cuda_op_topk_moe when weights overlap logits, so the kernel writes weights to the temp buffer, then copies them back after the kernel finishes on the same stream.

**Comment by am17an** [comment]
At the moment I don't think it's worth it to add an extra buffer for this. The fusion code should only activate when the conditions are right, rather than manipulating the conditions.
